File: java-app/src/main/java/com/showcase/java/domain/bus/EventBus.java
package com.showcase.java.domain.bus;

import com.showcase.java.domain.event.DomainEvent;

import java.util.List;
import java.util.function.Consumer;

public interface EventBus {
    void publish(List<DomainEvent> events);
    void subscribe(Consumer<DomainEvent> handler);
}

File: kotlin-app/src/main/kotlin/com/showcase/kotlin/domain/bus/EventBus.kt
package com.showcase.kotlin.domain.bus

import com.showcase.kotlin.domain.event.DomainEvent
import java.util.concurrent.CopyOnWriteArrayList

interface EventBus {
    suspend fun publish(events: List<DomainEvent>)
    fun subscribe(handler: suspend (DomainEvent) -> Unit)
}

class InMemoryEventBus : EventBus {
    private val handlers = CopyOnWriteArrayList<suspend (DomainEvent) -> Unit>()

    override suspend fun publish(events: List<DomainEvent>) {
        for (event in events) {
            for (handler in handlers) {
                handler(event)
            }
        }
    }

    override fun subscribe(handler: suspend (DomainEvent) -> Unit) {
        handlers.add(handler)
    }
}
